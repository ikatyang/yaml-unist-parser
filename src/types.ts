export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Node {
  type: string;
  position: Position;
}

export interface Parent extends Node {
  children: Node[];
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor: Null | Anchor;
  tag: Null | VerbatimTag | ShorthandTag | NonSpecificTag;
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
}

export interface CommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
  /** comments on the same line of the node */
  trailingComments: Comment[];
}

export interface EndCommentAttachable {
  /** comments after the node with greater column */
  endComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | VerbatimTag
  | ShorthandTag
  | NonSpecificTag
  | Anchor
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowMappingItem
  | FlowSequence
  | FlowSequenceItem
  | Null;

export type ContentNode =
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | Sequence
  | FlowMapping
  | FlowSequence
  | Null;

// -----------------------------------------------------------------------------

export interface Null extends Node {
  type: "null";
  /** always -1 */
  position: Position;
}

export interface Comment extends Node {
  type: "comment";
  value: string;
  /** @internal non-enumerable */
  parent?: Extract<YamlUnistNode, CommentAttachable>;
}

export interface Anchor extends Node {
  type: "anchor";
  value: string;
}

export interface VerbatimTag extends Node {
  type: "verbatimTag";
  value: string;
}

export interface ShorthandTag extends Node {
  type: "shorthandTag";
  handle: string;
  suffix: string;
}

export interface NonSpecificTag extends Node {
  type: "nonSpecificTag";
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent, CommentAttachable {
  type: "document";
  children: [DocumentHead, DocumentBody];
  /** always 0 */
  leadingComments: Comment[];
  /** only attachable on `...` */
  trailingComments: Comment[];
}

export interface DocumentHead extends Parent {
  type: "documentHead";
  children: Array<Comment | Directive>;
}

export interface DocumentBody extends Parent {
  type: "documentBody";
  children: Array<Comment | ContentNode>;
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Node, Content, CommentAttachable {
  type: "alias";
  value: string;
}

export interface BlockValue extends Node, Content, CommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent: null | number;
  value: string;
  /** comments between indicator and value */
  trailingComments: Comment[];
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Node, Content, CommentAttachable {
  type: "plain";
  value: string;
}

export interface QuoteValue extends Node, Content, CommentAttachable {
  value: string;
}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, CommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItemBase extends Parent, CommentAttachable {
  /** key-value pair */
  children: [MappingKey | Null, MappingValue | Null];
}

export interface MappingItem extends MappingItemBase {
  type: "mappingItem";
  children: [MappingKey, MappingValue | Null];
}

export interface MappingKey
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingKey";
  children: [ContentNode];
}

export interface MappingValue
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingValue";
  children: [ContentNode];
}

export interface Sequence
  extends Parent,
    Content,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItemBase extends Parent {
  children: [ContentNode];
}

export interface SequenceItem
  extends SequenceItemBase,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequenceItem";
}

export interface FlowCollection extends Parent, Content, CommentAttachable {
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: FlowMappingItem[];
}

export interface FlowMappingItem extends MappingItemBase {
  type: "flowMappingItem";
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowSequenceItem extends SequenceItemBase {
  type: "flowSequenceItem";
}
