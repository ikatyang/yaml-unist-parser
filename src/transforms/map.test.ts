import {
  getFirstContent,
  TestCase,
  testCases,
  testSyntaxError,
} from "../helpers";
import { Mapping, Root } from "../types";

testSyntaxError("asd: 123\nqwe\n");

testCases([
  ...createTestCases(":   "),
  ...createTestCases(": 123  "),
  ...createTestCases("a:   "),
  ...createTestCases("a: 123  "),
  ...createTestCases("a: 123  \nb: 456  ", true),
  ...createTestCases("? a  \n: 123  \n? b  \n: 456  ", true),
  ...createTestCases("? abc\n? def", true),
  ...createTestCases("? 123"),
  ...createTestCases("def:     456 # hello"),
  [
    "!!map\n #123 \n&anchor # 456\na: 123",
    [
      getMapping(),
      root => getFirstContent<Mapping>(root).middleComments[0],
      root => getFirstContent<Mapping>(root).middleComments[1],
    ],
  ],
]);

function createTestCases(text: string, hasSecondItem = false): TestCase[] {
  return !hasSecondItem
    ? [
        [text, getMapping()],
        [text, getMappingItem(0)],
        [text, getMappingKey(0)],
        [text, getMappingValue(0)],
      ]
    : [
        [text, getMapping()],
        [text, getMappingItem(0)],
        [text, getMappingItem(1)],
        [text, getMappingKey(0)],
        [text, getMappingKey(1)],
        [text, getMappingValue(0)],
        [text, getMappingValue(1)],
      ];
}

function getMapping() {
  return getFirstContent<Mapping>();
}

function getMappingItem(itemIndex: number) {
  return (root: Root) => getMapping()(root).children[itemIndex];
}

function getMappingKey(itemIndex: number) {
  return (root: Root) => getMappingItem(itemIndex)(root).children[0];
}

function getMappingValue(itemIndex: number) {
  return (root: Root) => getMappingItem(itemIndex)(root).children[1];
}
