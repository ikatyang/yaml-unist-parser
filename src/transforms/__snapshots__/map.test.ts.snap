// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 1`] = `
mapping (4:1 ~ 4:7)
1 | !!map¶
2 | ·#123·¶
3 | &anchor·#·456¶
4 | a:·123¶
  | ^^^^^^
<mapping>
  <shorthandTag handle="!!" suffix="map" />
  <anchor value="anchor" />
  <middleComment value="123 ">
  <middleComment value=" 456">
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 2`] = `
comment (2:2 ~ 2:7)
1 | !!map¶
2 | ·#123·¶
  |  ^^^^^
3 | &anchor·#·456¶
4 | a:·123¶
<comment value="123 " />
`;

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 3`] = `
comment (3:9 ~ 3:14)
1 | !!map¶
2 | ·#123·¶
3 | &anchor·#·456¶
  |         ^^^^^
4 | a:·123¶
<comment value=" 456" />
`;

exports[`":   " 1`] = `
mapping (1:1 ~ 1:2)
1 | :···¶
  | ^
<mapping>
  <mappingItem>
    <mappingKey>
      <null />
    </mappingKey>
    <mappingValue>
      <null />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`":   " 2`] = `
mappingItem (1:1 ~ 1:2)
1 | :···¶
  | ^
<mappingItem>
  <mappingKey>
    <null />
  </mappingKey>
  <mappingValue>
    <null />
  </mappingValue>
</mappingItem>
`;

exports[`":   " 3`] = `
mappingKey (1:1 ~ 1:1)
1 | :···¶
  | ~
<mappingKey>
  <null />
</mappingKey>
`;

exports[`":   " 4`] = `
mappingValue (1:1 ~ 1:2)
1 | :···¶
  | ^
<mappingValue>
  <null />
</mappingValue>
`;

exports[`": 123  " 1`] = `
mapping (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <null />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`": 123  " 2`] = `
mappingItem (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <null />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`": 123  " 3`] = `
mappingKey (1:1 ~ 1:1)
1 | :·123··¶
  | ~
<mappingKey>
  <null />
</mappingKey>
`;

exports[`": 123  " 4`] = `
mappingValue (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"? 123" 1`] = `
mapping (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="123" />
    </mappingKey>
    <null />
  </mappingItem>
</mapping>
`;

exports[`"? 123" 2`] = `
mappingItem (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="123" />
  </mappingKey>
  <null />
</mappingItem>
`;

exports[`"? 123" 3`] = `
mappingKey (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mappingKey>
  <plain value="123" />
</mappingKey>
`;

exports[`"? 123" 4`] = `
null (-1:-1 ~ -1:-1)
1 | ?·123¶
<null />
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 1`] = `
mapping (1:1 ~ 4:6)
1 | ?·a··¶
  | ^^^^^^
2 | :·123··¶
  | ^^^^^^^^
3 | ?·b··¶
  | ^^^^^^
4 | :·456··¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="b" />
    </mappingKey>
    <mappingValue>
      <plain value="456" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 2`] = `
mappingItem (1:1 ~ 2:6)
1 | ?·a··¶
  | ^^^^^^
2 | :·123··¶
  | ^^^^^
3 | ?·b··¶
4 | :·456··¶
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 3`] = `
mappingItem (3:1 ~ 4:6)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
  | ^^^^^^
4 | :·456··¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="b" />
  </mappingKey>
  <mappingValue>
    <plain value="456" />
  </mappingValue>
</mappingItem>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 4`] = `
mappingKey (1:1 ~ 1:4)
1 | ?·a··¶
  | ^^^
2 | :·123··¶
3 | ?·b··¶
4 | :·456··¶
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 5`] = `
mappingKey (3:1 ~ 3:4)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
  | ^^^
4 | :·456··¶
<mappingKey>
  <plain value="b" />
</mappingKey>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 6`] = `
mappingValue (2:1 ~ 2:6)
1 | ?·a··¶
2 | :·123··¶
  | ^^^^^
3 | ?·b··¶
4 | :·456··¶
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 7`] = `
mappingValue (4:1 ~ 4:6)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
4 | :·456··¶
  | ^^^^^
<mappingValue>
  <plain value="456" />
</mappingValue>
`;

exports[`"? abc\\n? def" 1`] = `
mapping (1:1 ~ 2:6)
1 | ?·abc¶
  | ^^^^^^
2 | ?·def¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="abc" />
    </mappingKey>
    <null />
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="def" />
    </mappingKey>
    <null />
  </mappingItem>
</mapping>
`;

exports[`"? abc\\n? def" 2`] = `
mappingItem (1:1 ~ 1:6)
1 | ?·abc¶
  | ^^^^^
2 | ?·def¶
<mappingItem>
  <mappingKey>
    <plain value="abc" />
  </mappingKey>
  <null />
</mappingItem>
`;

exports[`"? abc\\n? def" 3`] = `
mappingItem (2:1 ~ 2:6)
1 | ?·abc¶
2 | ?·def¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="def" />
  </mappingKey>
  <null />
</mappingItem>
`;

exports[`"? abc\\n? def" 4`] = `
mappingKey (1:1 ~ 1:6)
1 | ?·abc¶
  | ^^^^^
2 | ?·def¶
<mappingKey>
  <plain value="abc" />
</mappingKey>
`;

exports[`"? abc\\n? def" 5`] = `
mappingKey (2:1 ~ 2:6)
1 | ?·abc¶
2 | ?·def¶
  | ^^^^^
<mappingKey>
  <plain value="def" />
</mappingKey>
`;

exports[`"? abc\\n? def" 6`] = `
null (-1:-1 ~ -1:-1)
1 | ?·abc¶
2 | ?·def¶
<null />
`;

exports[`"? abc\\n? def" 7`] = `
null (-1:-1 ~ -1:-1)
1 | ?·abc¶
2 | ?·def¶
<null />
`;

exports[`"a:   " 1`] = `
mapping (1:1 ~ 1:3)
1 | a:···¶
  | ^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <null />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a:   " 2`] = `
mappingItem (1:1 ~ 1:3)
1 | a:···¶
  | ^^
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <null />
  </mappingValue>
</mappingItem>
`;

exports[`"a:   " 3`] = `
mappingKey (1:1 ~ 1:2)
1 | a:···¶
  | ^
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a:   " 4`] = `
mappingValue (1:2 ~ 1:3)
1 | a:···¶
  |  ^
<mappingValue>
  <null />
</mappingValue>
`;

exports[`"a: 123  " 1`] = `
mapping (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a: 123  " 2`] = `
mappingItem (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  " 3`] = `
mappingKey (1:1 ~ 1:2)
1 | a:·123··¶
  | ^
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a: 123  " 4`] = `
mappingValue (1:2 ~ 1:7)
1 | a:·123··¶
  |  ^^^^^
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"a: 123  \\nb: 456  " 1`] = `
mapping (1:1 ~ 2:7)
1 | a:·123··¶
  | ^^^^^^^^^
2 | b:·456··¶
  | ^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="b" />
    </mappingKey>
    <mappingValue>
      <plain value="456" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a: 123  \\nb: 456  " 2`] = `
mappingItem (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
2 | b:·456··¶
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  \\nb: 456  " 3`] = `
mappingItem (2:1 ~ 2:7)
1 | a:·123··¶
2 | b:·456··¶
  | ^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="b" />
  </mappingKey>
  <mappingValue>
    <plain value="456" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  \\nb: 456  " 4`] = `
mappingKey (1:1 ~ 1:2)
1 | a:·123··¶
  | ^
2 | b:·456··¶
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a: 123  \\nb: 456  " 5`] = `
mappingKey (2:1 ~ 2:2)
1 | a:·123··¶
2 | b:·456··¶
  | ^
<mappingKey>
  <plain value="b" />
</mappingKey>
`;

exports[`"a: 123  \\nb: 456  " 6`] = `
mappingValue (1:2 ~ 1:7)
1 | a:·123··¶
  |  ^^^^^
2 | b:·456··¶
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"a: 123  \\nb: 456  " 7`] = `
mappingValue (2:2 ~ 2:7)
1 | a:·123··¶
2 | b:·456··¶
  |  ^^^^^
<mappingValue>
  <plain value="456" />
</mappingValue>
`;

exports[`"def:     456 # hello" 1`] = `
mapping (1:1 ~ 1:21)
1 | def:·····456·#·hello¶
  | ^^^^^^^^^^^^^^^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="def" />
    </mappingKey>
    <mappingValue>
      <plain value="456">
        <trailingComment value=" hello">
      </plain>
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"def:     456 # hello" 2`] = `
mappingItem (1:1 ~ 1:21)
1 | def:·····456·#·hello¶
  | ^^^^^^^^^^^^^^^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="def" />
  </mappingKey>
  <mappingValue>
    <plain value="456">
      <trailingComment value=" hello">
    </plain>
  </mappingValue>
</mappingItem>
`;

exports[`"def:     456 # hello" 3`] = `
mappingKey (1:1 ~ 1:4)
1 | def:·····456·#·hello¶
  | ^^^
<mappingKey>
  <plain value="def" />
</mappingKey>
`;

exports[`"def:     456 # hello" 4`] = `
mappingValue (1:4 ~ 1:21)
1 | def:·····456·#·hello¶
  |    ^^^^^^^^^^^^^^^^^
<mappingValue>
  <plain value="456">
    <trailingComment value=" hello">
  </plain>
</mappingValue>
`;

exports[`Implicit map keys need to be followed by map values 1`] = `
"Implicit map keys need to be followed by map values
1 | asd:·123¶
2 | qwe¶
  | ^^^^
3 | ¶
  | ^"
`;
